### DISCUSSION TITLE
Question about setting the endpoint url

### DISCUSSION THREAD

**Discussion by dubonzi** [open-telemetry/opentelemetry-go#3726] [Q&A] [answered]
Hi, I've been having some problems setting up the http exporter endpoint. The problem happens when I try to use an IP address as the endpoint.

Whenever I set my endpoint to '127.0.0.1:4318', I get the following error during tracer startup: 

`"msg"="parse url" "error"="parse \"127.0.0.1:4318\": first path segment in URL cannot contain colon" "input"="127.0.0.1:4318"`

I traced the error to [this point ](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/internal/envconfig/envconfig.go#L101).

If I try to use 'http://127.0.0.1:4318' instead, the url ends up being modified by the SDK and I get an error when it tries to export traces (not on startup):

`traces export: parse "http://http:%2F%2F127.0.0.1:4318/v1/traces": invalid URL escape "%2F"`

Seems like the schema gets duplicated and ends up creating an invalid URL, which is happening [here](https://github.dev/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracehttp/client.go#L207).

Im starting the exporter with these options:

```
otlptracehttp.WithEndpoint(endpoint),
otlptracehttp.WithInsecure()
```

Am I forgetting something?

**Comment by dmathieu** [comment]
I wonder if this is a parsing error due to using an IP address. What if you replace the IP with `localhost`?

**Comment by dubonzi** [comment]
Using `localhost:4318` works, but `http://localhost:4318/` produces the same error. In production we use an IP as endpoint, so thats why I'm testing with `127.0.0.1`. Seems like the schema should be set before anything else and the url shouldn't be modified later like happens in the second case.

**Reply by dmathieu** [reply]
> Using localhost:4318 works, but http://localhost:4318/ produces the same error. 

You should not set the protocol when using `WithEndpoint`. So it's expected that `http://localhost` won't work.

So, I have been trying this out with the namedtracer example: https://github.com/open-telemetry/opentelemetry-go/blob/main/example/namedtracer/main.go

Replacing the stdoutexporter with otlp:

```golang
func initTracer() error {
	exp, err := otlptracehttp.New(context.Background(),
		otlptracehttp.WithEndpoint("127.0.0.1:4318"),
	)

	if err != nil {
		return fmt.Errorf("failed to initialize stdouttrace exporter: %w", err)
	}
	bsp := sdktrace.NewBatchSpanProcessor(exp)
	tp = sdktrace.NewTracerProvider(
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
		sdktrace.WithSpanProcessor(bsp),
	)
	otel.SetTracerProvider(tp)
	return nil
}
```

And I'm not getting any parsing error, either using the HTTP exporter, or the gRPC one.
I don't have anything running on that port, so I'm getting a connection refused error, but that's expected.

**Comment by dubonzi** [comment, marked as answer]
Your example worked but it broke when I passed the endpoint as an environment variable `OTEL_EXPORTER_OTLP_ENDPOINT=127.0.0.1:4318`, can you try it like that? Apparently the variable is taking precedence over `WithEndpoint` in this case.

Looking at the code, the env var is [parsed as an URL](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/internal/envconfig/envconfig.go#L101) while in `WithEndpoint` the value is [assigned directly as a string](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/internal/otlpconfig/options.go#L246) without the parse.

**Reply by dmathieu** [reply]
True, the environment variable requires a protocol. The following works:
```
OTEL_EXPORTER_OTLP_ENDPOINT=http://127.0.0.1:4318 go run main.go
```

This behavior matches the specification requirement though: https://github.com/open-telemetry/opentelemetry-specification/blob/102b00daef366e48d99f672f343fe78f777fd3af/specification/protocol/exporter.md?plain=1#L11

I wonder if we should do the same for `WithEndpoint`, so it matches the environment variable.

**Reply by dubonzi** [reply]
Humm I see, so I shouldn't use `WithEndpoint` while the environment variable is set. What I'm doing is using the same value from the environment variable in `WithEndpoint`, so I end up in a scenario where one breaks if the url has a schema, the other if it doesn't.

> I wonder if we should do the same for WithEndpoint, so it matches the environment variable.

I think that would be ideal to keep the behaviour consistent.

**Reply by dmathieu** [reply]
See this issue: https://github.com/open-telemetry/opentelemetry-go/issues/3730

**Reply by dubonzi** [reply]
Thanks! I'll mark this as answered and follow the issue for updates.
